### ISSUE TITLE
First-Class Windows Development Environment

### ISSUE THREAD

**Issue by jdegoes** [golemcloud/golem#1210]
In the Golem project currently, Windows is not a first-class development experience. The problems are many and range from:

1. How Golem is built and tested
2. Which Rust libraries are used, and to what degree they support Windows
3. Random scripts and tools that may not support Windows fully or at all

This ticket is to enhance Golem so that developers who are using Windows may have a first-class development experience, without having to install Windows Subsystem for Linux (WSL). 

When completed, a clean checkout of the Golem code base will:

1. Fully build
2. Reliably pass all tests
3. Require no more setup than development on Linux or Mac

Moreover, Github-based CI will automatically run tests on Windows to ensure no future regressions.

**Comment by zelosleone** [comment]
Truth to be told the major problem is wasmtime package versioning making it hard to build-test, linking version is extremely hard to pass unless default features of wasmtime is completely disabled, which could make it unable to connect on HTTP via python. But there are also at least 4 years old crates like poule that is not updated recently, I am pretty sure some golem repo code could be changed to fit the newer/more maintained crates.

**Comment by zelosleone** [comment]
> @zelosleone The issue of conflicting wasm appearing in both Windows and Linux after the merge of #1200. Before it worked fine without any conflicts of wasm versions. The problem stems from wasm-wave likely pulling older version of wasm as its dependency



Nope, its the wib-parser that pulls the old versioning, which breaks the wib module of the golem if upgraded. 0.6.0 is p old for wib-parser crate since it went more than 5 updates from that time, so golem code needs to change and be tested that "Yeah, this proves upgrading it didnt break anything after we updated the code and it works like old"

**Comment by zelosleone** [comment]
> @zelosleone The wit parser conflict, I think not under consideration of this issue. It happening with other os linux too not sure what about Mac though.
> 
> I have clearly discussed with jdegoes and vigoo for the compatibility of windows and the alternatives and following them to solve keeping in mind of future work of Golem

Yeah I was working for 10 days on windows with golem so i had a headstart, thats fine, I almost finished the PR.

**Comment by Aditya-Choudhry** [comment]
Hi, I would like to work on this issue. Could you assign it to me?

**Comment by parvit** [comment]
/attempt #1210 

@jdegoes Do you require for the second point of eliminating docker/WSL requirement to provide also scripts for setup of services (eg. postgres) or should this change assume the services already installed on the test machine ? thank you.

**Comment by laz-001** [comment]
@jdegoes , would you consider splitting this into multiple bounties?

I am personally interested that golem is 'first-class' on windows systems.

I could work on this, to split this down to multiple issues (which can then get their own bounties).

Be aware that this needs most possibly changes to the main code (code, libraries, ci) - all should be doable though and increase the general code/project quality.

**Comment by laz-001** [comment]
@Nanashi-lab , why would you 'thumb-down' this? Feels a bit like kinder-garden.

Just take a look at the history of this issue, and the original text.

It is clear (at least for higher-XP devs) that this needs to be broken down.

**Comment by parvit** [comment]
> [@jdegoes](https://github.com/jdegoes)
> 
> > 2. Reliably pass all tests locally without requiring docker or WSL
> 
> The ci tests depend on docker, e.g. launching postgres via docker.
> 
> [@parvit](https://github.com/parvit)
> 
> > Do you require for the second point of eliminating docker/WSL requirement
> 
> be aware of the above

You seem to have misinterpreted my comment, what i meant is that to make the change to not require docker anymore (not used only in ci but also locally), you can either:

1. require to have services installed and setup by the user

2. install and setup the services via a script in the project

The tests would then operate by only calling the services like at the moment and reset the state inbetween. 

The choice is for @jdegoes to make but currently i'm working on the other points so there's no rush on my part.

**Comment by laz-001** [comment]
> The choice is for [@jdegoes](https://github.com/jdegoes) to make but currently i'm working on the other points so there's no rush on my part, however you .

@jdegoes is (quite obviously) overworked. Response times here and in other issues are low.

And he should not bother with such tiny details "scripts or not" (though he should be  bothered with refining the requirements a bit, and of course with splitting down this issue/bounty or extending it with a management-issue/bounty at ~$1K)

Thing is:
* removing WSL dependency: required
* removing docker dependency: NOT required (the opposite, devs should be able to simply use docker, if prefered, otherwise go with an installed postgres.

I've recently installed postgres on Win10, it quite trivial, not much to script there.

> , however you 

?

(I comment often on counties, just to make then reach fulfillment sooner. This does not mean always that I want to take the bounty over. Mostly I just assist colleagues - we bounty-workers are kind of colleagues).

**Comment by parvit** [comment]
> > The choice is for [@jdegoes](https://github.com/jdegoes) to make but currently i'm working on the other points so there's no rush on my part, however you .
> 
> [@jdegoes](https://github.com/jdegoes) is (quite obviously) overworked. Response times here and in other issues are low.
> 
> And he should not bother with such tiny details "scripts or not" (though he should be bothered with refining the requirements a bit, and of course with splitting down this issue/bounty or extending it with a management-issue/bounty at ~$1K)
> 
> Thing is:
> 
> * removing WSL dependency: required
> * removing docker dependency: NOT required (the opposite, devs should be able to simply use docker, if prefered, otherwise go with an installed postgres.
> 
> I've recently installed postgres on Win10, it quite trivial, not much to script there.
> 
> > , however you
> 
> ?
> 
> (I comment often on counties, just to make then reach fulfillment sooner. This does not mean always that I want to take the bounty over. Mostly I just assist colleagues - we bounty-workers are kind of colleagues).

Sure you may see the matter however you like and implement accordingly, i'll do the same myself.

**Comment by parvit** [comment]
@jdegoes so we can assume the service dependencies (tests do interact with postgres/minio) can be left to the user and only the configuration of the tests are to be part of the test (eg. username/password/hostname etc)?

**Comment by jdegoes** [comment]
I want `cargo build` and `cargo test` to work immediately after checkout on any Windows machine, without the developer having to do anything at all other than type these commands after checkout.

**Comment by parvit** [comment]
Would it satisfy your requirement if the tests that right now require interaction with docker containers instead use mock calls so its not necessary anymore to install and configure anything ? I think that would be an improvement also for the other platforms.

**Comment by Rutik7066** [comment]
Yes, all tools needed for Golem development can be installed on Windows. Some tools, like `lnav` and `redis`, don’t officially support Windows, but we can use the MSYS2 builds for that. Currently, the Windows build for `lnav` isn’t available in the official releases. I’m helping test `lnav` on Windows, and once it makes it into the official releases, I’ll automate its installation in a script.

That script will be added to the `dev-flow` or `build` job in `Makefile.toml`, so developers won’t need to do any manual setup—it’ll be fully automated.

Previously, we had issues building Golem tests, but that was due to an outdated `wasmtime` version. The problem was resolved in `wasmtime` v27. I flagged this on Discord to @vigoo, and we’ve now updated to the latest major version.

Tests are building perfectly now. I’ve run them, and they’re all passing. I ran them one by one, though, because my laptop (`i5 CPU`, `16GB RAM`) couldn’t handle running `cargo make tests` all at once.

**Comment by Rutik7066** [comment]
@parvit  Looks like I am missing something. I don't understand why we need Docker when running tests.

**Comment by parvit** [comment]
@Rutik7066 An example: https://github.com/golemcloud/golem/blob/2b0dd21bf1461d59c9353e0a84b0f825d222056f/golem-service-base/tests/blob_storage.rs#L111-L119

This spins up a MinIO container for the test and same is done in other tests for postgres etc. Removing docker requirement means to fix these tests to work without it. I assume all tests unit and integration are required to be updated and pass for this request.

**Comment by Rutik7066** [comment]
@parvit Thank you so much for pointing this out. We can work together on this. What say?

**Comment by laz-001** [comment]
@jdegoes, you MUST be more precise/thorough with your requirements.

> I want `cargo build` and `cargo test` to work immediately after checkout on any Windows machine

=> System Requirements

"any Windows machine", => not feasible (You need to specify the win requirements, e.g. Win 10, Win 10 pro with docker)

> , without the developer having to do anything at all other than type these commands after checkout.

Again, not feasible.

=> Prerequisites (setup cargo, docker etc.)

> Neither Docker nor WSL should be necessary.

The linux tests depend on Docker.

Quite naturally, the Windows executed tests will depend on Docker, too.

That's why I posted this comment: https://github.com/golemcloud/golem/issues/1210#issuecomment-2839833661

It is clear to me that you're a bit overworked, but you MUST ensure to invest some time validate the requirements first.

No time? Add a bounty for this.

**Comment by parvit** [comment]
> [@parvit](https://github.com/parvit) Thank you so much for pointing this out. We can work together on this. What say?

I'd like first to have confirmation on how to proceed for removing the wsl/docker dependency, that integration tests need to use mocks or just services local to the machine.

**Comment by laz-001** [comment]
@jdegoes , this is looking very good. The 'drama' is (luckily) not with rust-crates, but with different virtualization behavior/problems/limits (e.g. github windows runners do not support `linux/amd64`).

I should have on Monday/Tuesday the usable midway-result (a windows-ci draft that gives feedback to developers), then maybe another week to stabilize/document everything production-grade.

(local development will be MAYBE possible without WSL/HyperV, but there are practical limits, the 'whole world' works `docker/linux, usually amd64`).

### Question 

Could we split the bounty in two parts (40/60)?

Mainly because the one planned week can easily become 2 weeks.

**Comment by jdegoes** [comment]
@laz-001 This feature has no value to me unless it is completed in its entirety.

**Comment by laz-001** [comment]
> This feature has no value to me unless it is completed in its entirety.

This isn't a feature, it's a system-wide update.

9 persons started and dropped it - didn't you wonder why?

I sliced this system-update (at my 2nd attempt) into manageable chunks, starting with a CI run (which I have ready, but not PR'd)

If you can't see the benefit, the value (in itself) and the necessity of a working CI run, then we have very different views of "System Design".

Mine are trained from hardware-design, where you simply cannot survive without splitting bigger tasks down into chunks.

Anyways, given the project-organization, the open (and risky) bounties, the fuzzy requirements, the non-reply to (justified) questions re requirements and some more things, I decided to back off, instead of finalizing and providing the CI result.

It's just too risky that I loose even more time.

If you're willing to increase the budget to $10K, split it down to 4 chunks, then I'd take this on again (potentially isolating sub-tasks and preparing them, thus other devs can solve and claim).

* A: CI
* B: Local Build & Tests
* C: Fixes & Workarounds for failing subsystems
* D: Production Grade Refinements

**Comment by parvit** [comment]
> > This feature has no value to me unless it is completed in its entirety.
> 
> This isn't a feature, it's a system-wide update.
> 
> 9 persons started and dropped it - didn't you wonder why?
> 
> I sliced this system-update (at my 2nd attempt) into manageable chunks, starting with a CI run (which I have ready, but not PR'd)
> 
> If you can't see the benefit, the value (in itself) and the necessity of a working CI run, then we have very different views of "System Design".
> 
> Mine are trained from hardware-design, where you simply cannot survive without splitting bigger tasks down into chunks.
> 
> Anyways, given the project-organization, the open (and risky) bounties, the fuzzy requirements, the non-reply to (justified) questions re requirements and some more things, I decided to back off, instead of finalizing and providing the CI result.
> 
> It's just too risky that I loose even more time.
> 
> If you're willing to increase the budget to $10K, split it down to 4 chunks, then I'd take this on again (potentially isolating sub-tasks and preparing them, thus other devs can solve and claim).
> 
> * A: CI
> * B: Local Build & Tests
> * C: Fixes & Workarounds for failing subsystems
> * D: Production Grade Refinements

Can't speak for the others but i surely have *not* dropped this. In fact i've nearly finished fixing all the unit and integration tests as requested by separating the docker requirement (it's not ready yet but i'll share my actual branch with instructions at a later time).

CI is the next step as written in the ticket description.

**Comment by jdegoes** [comment]
> 9 persons started and dropped it - didn't you wonder why?

I honestly don't care. If you aren't up for the requirements, don't take it on. 😉

**Comment by rutikthakre** [comment]
Dude your PR solve the problem which is not the problem. 
> > I honestly don't care. 
> 
> It's obvious that you don't care - the current code/project quality standards reflect this clearly.
> 
> > If you aren't up
> 
> ???
> 
> This is **trivial grinding work** for devs of my level.
> 
> >  for the requirements, 
> 
> Requirements? Those are not "requirements" but unfounded/inconsistent wishes:
> 
> Trying to clarify/derive real requirements is not possible.
> 
> > don't take it on.
> 
> Here you have a point.
> 
> Should simply listen to my intuition more...
> 
> Have fun, puppet-master.
> 
> .

**Comment by vigoo** [comment]
We are closing this bounty as it did not reach its original goal.

**Comment by parvit** [comment]
Please reconsider! I've even sent a PR already for part of the solution here https://github.com/golemcloud/moonbit-component-generator/pull/7 !

**Comment by parvit** [comment]
I was not ready yet to publish the PR but i've created it hoping to continue the work if this is reopened.
